rustsec/src/repository/git/repository.rs

```diff
@@ -72,64 +69,99 @@ impl Repository {
-                let remote_target = remote_main_ref.target().unwrap();
-
-                // Set the local main ref to match the remote
-                match repo.find_reference(LOCAL_REF) {
-                    Ok(mut local_main_ref) => {
-                        local_main_ref.set_target(
-                            remote_target,
-                            &format!(
-                                "rustsec: moving `main` to {}: {}",
-                                REMOTE_REF, &remote_target
-                            ),
-                        )?;
-                    }
-                    Err(e) if e.code() == git2::ErrorCode::NotFound => {
-                        // TODO(tarcieri): remove this workaround after repos have migrated
-                        let old_ref = repo.find_reference("refs/heads/master")?;
-                        git2::Branch::wrap(old_ref).rename("main", true)?;
-                        Self::fetch(url, &path, ensure_fresh)?;
-                    }
-                    Err(e) => {
-                        return Err(e.into());
-                    }
-                };
+            let res = if let Some(repo) = repo {
+                (repo, None)
             } else {
-                git2::build::RepoBuilder::new()
-                    .fetch_options(fetch_opts)
-                    .clone(url, &path)?;
-            }
+                let mut progress = gix::progress::Discard;
+                let should_interrupt = &gix::interrupt::IS_INTERRUPTED;
 
-            Ok(())
-        })?;
+                let (mut prep_checkout, out) = gix::prepare_clone(url, path)
+                    .map_err(|err| {
+                        format_err!(ErrorKind::Repo, "failed to prepare clone: {}", err)
+                    })?
+                    .with_remote_name("origin")
+                    .map_err(|err| format_err!(ErrorKind::Repo, "invalid remote name: {}", err))?
+                    .configure_remote(|remote| Ok(remote.with_refspecs([REF_SPEC], DIR)?))
+                    .fetch_then_checkout(&mut progress, should_interrupt)
+                    .map_err(|err| format_err!(ErrorKind::Repo, "failed to fetch repo: {}", err))?;
+
+                let repo = prep_checkout
+                    .main_worktree(&mut progress, should_interrupt)
+                    .map_err(|err| {
+                        format_err!(ErrorKind::Repo, "failed to checkout fresh clone: {}", err)
+                    })?
+                    .0;
+
+                (repo, Some(out))
+            };
 
-        let repo = Self::open(path)?;
-        let latest_commit = repo.latest_commit()?;
+            Ok(res)
+        };
+
+        let (mut repo, fetch_outcome) = open_or_clone_repo()?;
+
+        if let Some(fetch_outcome) = fetch_outcome {
+            tame_index::utils::git::write_fetch_head(
+                &repo,
+                &fetch_outcome,
+                &repo.find_remote("origin").unwrap(),
+            )?;
+        } else {
```

Comment:
Note that in both cases we write a `FETCH_HEAD` (gix doesn't yet do this). IMO this should be the timestamp used for staleness detection, but since I wanted to do a straight port it still uses the commit time of the HEAD commit.

Reply:
Keeping it as-is is a good call. Relying on `FETCH_HEAD` would introduce issues around relying on the client time - e.g. if the client time is set into the far future at the time of checkout, the DB would never be re-downloaded again.

There are of course possible mitigations to this but they add a lot more complexity.